src/nvim/ex_docmd.c

```diff
@@ -5600,6 +5600,32 @@ static void ex_detach(exarg_T *eap)
   }
 }
 
+/// ":connect"
+///
+/// Connects the current UI to a different server
+///
+/// ":connect <address>" detaches the current UI and connects to the given server.
+/// ":connect! <address>" stops the current server if no other UIs are attached, then connects to the given server.
+static void ex_connect(exarg_T *eap)
+{
+  bool stop_server = eap->forceit ? (ui_active() == 1) : false;
+
+  Error err = ERROR_INIT;
+  remote_ui_connect(current_ui, eap->arg, &err);
+
+  if (ERROR_SET(&err)) {
+    emsg(err.msg);
+    api_clear_error(&err);
+    return;
+  }
+
+  ex_detach(NULL);
+  if (stop_server) {
+    exiting = true;
+    getout(0);
```

Comment:
This doesn't seem like the right approach. For `:restart`, the UI client calls `:qall!` on the server. That might be more reliable, and avoid the ASAN issue. It keeps the decisions all in one place (the client), and avoids race conditions.

Was there a reason for doing this this way?